lib/onig-lib.js

```diff
@@ -0,0 +1,184 @@
+/**
+ * @import {IOnigMatch, OnigScanner, OnigString} from 'vscode-textmate'
+ */
+
+import {toRegExp} from 'oniguruma-to-es'
+
+/** @type {boolean} */
+const forgiving = true
+
+const MAX = 2 ** 32 - 1
+
+/**
+ * Create an oniguruma scanner compatible with `vscode-textmate`.
+ *
+ * This is based on https://github.com/shikijs/shiki/blob/main/packages/engine-javascript/src/scanner.ts
+ *
+ * @param {Array<string>} sources
+ *   The regular expressions to compile.
+ * @returns {OnigScanner}
+ *   The oniguruma scanner.
+ */
+export function createOnigScanner(sources) {
+  /** @type {Map<string, RegExp | Error>} */
+  const cache = new Map()
+  const regexps = sources.map((pattern) => {
+    if (typeof pattern !== 'string') {
+      return pattern
+    }
+
+    // Cache
+    const cached = cache.get(pattern)
+    if (cached) {
+      if (cached instanceof RegExp) {
+        return cached
+      }
+
+      if (forgiving) {
+        return null
+      }
+
+      throw cached
+    }
+
+    try {
+      const regex = toRegExp(pattern, {
+        global: true,
+        hasIndices: true,
+        // This has no benefit for the standard JS engine, but it avoids a perf penalty for
+        // precompiled grammars when constructing extremely long patterns that aren't always used
+        lazyCompileLength: 3000,
+        rules: {
+          // Needed since TextMate grammars merge backrefs across patterns
+          allowOrphanBackrefs: true,
+          // Improves search performance for generated regexes
+          asciiWordBoundaries: true,
+          // Follow `vscode-oniguruma` which enables this Oniguruma option by default
+          captureGroup: true,
+          // Oniguruma uses depth limit `20`; lowered here to keep regexes shorter and maybe
+          // sometimes faster, but can be increased if issues reported due to low limit
+          recursionLimit: 5,
+          // Oniguruma option for `^`->`\A`, `$`->`\Z`; improves search performance without any
+          // change in meaning since TM grammars search line by line
+          singleline: true
+        }
+      })
+      cache.set(pattern, regex)
+      return regex
+    } catch (error_) {
```

Comment:
I’d call this `cause`?

Reply:
Me too, but XO insisted I call this `error_`.

If we decide to not be forgiving, there‘s no point in caching errors IMO. In that case I will remove the entire try/catch.